OrbitQt/StatusListenerImpl.cpp

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "StatusListenerImpl.h"
+
+#include "OrbitBase/Logging.h"
+
+uint64_t StatusListenerImpl::AddStatus(std::string message) {
+  uint64_t id = GetNextId();
+  status_bar_->showMessage(QString::fromStdString(message));
+
+  status_messages_.insert_or_assign(id, std::move(message));
+  status_id_stack_.push_back(id);
+
+  return id;
+}
+
+void StatusListenerImpl::ClearStatus(uint64_t status_id) {
+  CHECK(status_messages_.count(status_id) == 1);
+  status_messages_.erase(status_id);
+  status_id_stack_.erase(std::remove(status_id_stack_.begin(), status_id_stack_.end(), status_id),
+                         status_id_stack_.end());
+  if (status_id_stack_.empty()) {
+    status_bar_->clearMessage();
+    return;
+  }
+
+  uint64_t current_status_id = status_id_stack_.back();
+  status_bar_->showMessage(QString::fromStdString(status_messages_.at(current_status_id)));
+}
+
+void StatusListenerImpl::UpdateStatus(uint64_t status_id, std::string message) {
+  CHECK(status_messages_.count(status_id) == 1);
+  auto it = std::find(status_id_stack_.begin(), status_id_stack_.end(), status_id);
+  CHECK(it != status_id_stack_.end());
+
+  status_id_stack_.erase(it);
+  status_id_stack_.push_back(status_id);
+
+  status_bar_->showMessage(QString::fromStdString(message));
+  status_messages_.insert_or_assign(status_id, std::move(message));
+}
+
+std::unique_ptr<StatusListener> StatusListenerImpl::Create(QStatusBar* status_bar) {
+  return std::unique_ptr<StatusListener>(new StatusListenerImpl(status_bar));
+}
+
+uint64_t StatusListenerImpl::GetNextId() {
+  uint64_t current_id = next_id_++;
+  uint64_t candidate = current_id;
+  do {
+    if (status_messages_.count(candidate) == 0) {
+      return candidate;
+    }
+    candidate = next_id_++;
+  } while (candidate != current_id);
+
+  FATAL("Failed to find free (64bit) id and somehow we did not run out of memory before that.");
+}
```

Comment:
Whoa. If I understand correctly, this is to handle a uint64_t value wrapping around? That's an astronomical number of status updates, it is safe to say that this will never happen :-). At 1GHz updates, it would take over 500 years. I'd be happy with a simple "next_id_++".

Reply:
ok, I simplified this.